File: runtime/Cpp/org/antlr/v4/runtime/ListTokenSource.cpp
#include "ListTokenSource.h"
#include "Token.h"

/*
 * [The "BSD license"]
 *  Copyright (c) 2013 Terence Parr
 *  Copyright (c) 2013 Dan McLaughlin
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. The name of the author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

namespace org {
    namespace antlr {
        namespace v4 {
            namespace runtime {

                int ListTokenSource::getCharPositionInLine() {
                    if (i < tokens.size()) {
                        return ((Token*)tokens[i])->getCharPositionInLine();
                    } else if (eofToken != nullptr) {
                        return eofToken->getCharPositionInLine();
                    } else if (tokens.size() > 0) {
                        // have to calculate the result from the line/column of the previous
                        // token, along with the text of the token.
                        Token *lastToken = tokens[tokens.size() - 1];
                        std::wstring tokenText = lastToken->getText();
                        if (tokenText != L"") {
                            int lastNewLine = (int)tokenText.rfind(L'\n');
                            if (lastNewLine >= 0) {
                                return (int)tokenText.length() - lastNewLine - 1;
                            }
                        }

                        return lastToken->getCharPositionInLine() + lastToken->getStopIndex() - lastToken->getStartIndex() + 1;
                    }

                    // only reach this if tokens is empty, meaning EOF occurs at the first
                    // position in the input
                    return 0;
                }

                    Token *ListTokenSource::nextToken() {
                    if (i >= tokens.size()) {
                        if (eofToken == nullptr) {
                            int start = -1;
                            if (tokens.size() > 0) {
                                int previousStop = ((Token*)tokens[tokens.size() - 1])->getStopIndex();
                                if (previousStop != -1) {
                                    start = previousStop + 1;
                                }
                            }

                            int stop = std::max(-1, start - 1);
                            eofToken = _factory->create(new std::pair<TokenSource*, CharStream*>(this, getInputStream()), Token::_EOF, L"EOF", Token::DEFAULT_CHANNEL, start, stop, getLine(), getCharPositionInLine());
                        }

                        return eofToken;
                    }

                    Token *t = tokens[i];
                    if (i == tokens.size() - 1 && t->getType() == Token::_EOF) {
                        eofToken = t;
                    }

                    i++;
                    return t;
                }

                int ListTokenSource::getLine() {
                    if (i < tokens.size()) {
                        return tokens[i]->getLine();
                    } else if (eofToken != nullptr) {
                        return eofToken->getLine();
                    } else if (tokens.size() > 0) {
                        // have to calculate the result from the line/column of the previous
                        // token, along with the text of the token.
                        Token *lastToken = tokens[tokens.size() - 1];
                        int line = lastToken->getLine();

                        std::wstring tokenText = lastToken->getText();
                        if (tokenText != L"") {
                            for (size_t i = 0; i < tokenText.length(); i++) {
                                if (tokenText[i] == L'\n') {
                                    line++;
                                }
                            }
                        }

                        // if no text is available, assume the token did not contain any newline characters.
                        return line;
                    }

                    // only reach this if tokens is empty, meaning EOF occurs at the first
                    // position in the input
                    return 1;
                }

                CharStream *ListTokenSource::getInputStream() {
                    if (i < tokens.size()) {
                        return tokens[i]->getInputStream();
                    } else if (eofToken != nullptr) {
                        return eofToken->getInputStream();
                    } else if (tokens.size() > 0) {
                        return tokens[tokens.size() - 1]->getInputStream();
                    }

                    // no input stream information is available
                    return nullptr;
                }

                std::string ListTokenSource::getSourceName() {
                    if (sourceName != "") {
                        return sourceName;
                    }

                    CharStream *inputStream = getInputStream();
                    if (inputStream != nullptr) {
                        return inputStream->getSourceName();
                    }

                    return "List";
                }

                TokenFactory<CommonToken*> *ListTokenSource::getTokenFactory() {
                    return _factory;
                }

                void ListTokenSource::InitializeInstanceFields() {
                    i = 0;
                    _factory = CommonTokenFactory::DEFAULT;
                }
            }
        }
    }
}

File: runtime/Java/src/org/antlr/v4/runtime/ListTokenSource.java
/*
 * Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

package org.antlr.v4.runtime;

import org.antlr.v4.runtime.misc.Pair;

import java.util.List;

/**
 * Provides an implementation of {@link TokenSource} as a wrapper around a list
 * of {@link Token} objects.
 *
 * <p>If the final token in the list is an {@link Token#EOF} token, it will be used
 * as the EOF token for every call to {@link #nextToken} after the end of the
 * list is reached. Otherwise, an EOF token will be created.</p>
 */
public class ListTokenSource implements TokenSource {
	/**
	 * The wrapped collection of {@link Token} objects to return.
	 */
	protected final List<? extends Token> tokens;

	/**
	 * The name of the input source. If this value is {@code null}, a call to
	 * {@link #getSourceName} should return the source name used to create the
	 * the next token in {@link #tokens} (or the previous token if the end of
	 * the input has been reached).
	 */
	private final String sourceName;

	/**
	 * The index into {@link #tokens} of token to return by the next call to
	 * {@link #nextToken}. The end of the input is indicated by this value
	 * being greater than or equal to the number of items in {@link #tokens}.
	 */
	protected int i;

	/**
	 * This field caches the EOF token for the token source.
	 */
	protected Token eofToken;

	/**
	 * This is the backing field for {@link #getTokenFactory} and
	 * {@link setTokenFactory}.
	 */
	private TokenFactory<?> _factory = CommonTokenFactory.DEFAULT;

	/**
	 * Constructs a new {@link ListTokenSource} instance from the specified
	 * collection of {@link Token} objects.
	 *
	 * @param tokens The collection of {@link Token} objects to provide as a
	 * {@link TokenSource}.
	 * @exception NullPointerException if {@code tokens} is {@code null}
	 */
	public ListTokenSource(List<? extends Token> tokens) {
		this(tokens, null);
	}

	/**
	 * Constructs a new {@link ListTokenSource} instance from the specified
	 * collection of {@link Token} objects and source name.
	 *
	 * @param tokens The collection of {@link Token} objects to provide as a
	 * {@link TokenSource}.
	 * @param sourceName The name of the {@link TokenSource}. If this value is
	 * {@code null}, {@link #getSourceName} will attempt to infer the name from
	 * the next {@link Token} (or the previous token if the end of the input has
	 * been reached).
	 *
	 * @exception NullPointerException if {@code tokens} is {@code null}
	 */
	public ListTokenSource(List<? extends Token> tokens, String sourceName) {
		if (tokens == null) {
			throw new NullPointerException("tokens cannot be null");
		}

		this.tokens = tokens;
		this.sourceName = sourceName;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public int getCharPositionInLine() {
		if (i < tokens.size()) {
			return tokens.get(i).getCharPositionInLine();
		}
		else if (eofToken != null) {
			return eofToken.getCharPositionInLine();
		}
		else if (tokens.size() > 0) {
			// have to calculate the result from the line/column of the previous
			// token, along with the text of the token.
			Token lastToken = tokens.get(tokens.size() - 1);
			String tokenText = lastToken.getText();
			if (tokenText != null) {
				int lastNewLine = tokenText.lastIndexOf('\n');
				if (lastNewLine >= 0) {
					return tokenText.length() - lastNewLine - 1;
				}
			}

			return lastToken.getCharPositionInLine() + lastToken.getStopIndex() - lastToken.getStartIndex() + 1;
		}

		// only reach this if tokens is empty, meaning EOF occurs at the first
		// position in the input
		return 0;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Token nextToken() {
		if (i >= tokens.size()) {
			if (eofToken == null) {
				int start = -1;
				if (tokens.size() > 0) {
					int previousStop = tokens.get(tokens.size() - 1).getStopIndex();
					if (previousStop != -1) {
						start = previousStop + 1;
					}
				}

				int stop = Math.max(-1, start - 1);
				eofToken = _factory.create(new Pair<TokenSource, CharStream>(this, getInputStream()), Token.EOF, "EOF", Token.DEFAULT_CHANNEL, start, stop, getLine(), getCharPositionInLine());
			}

			return eofToken;
		}

		Token t = tokens.get(i);
		if (i == tokens.size() - 1 && t.getType() == Token.EOF) {
			eofToken = t;
		}

		i++;
		return t;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public int getLine() {
		if (i < tokens.size()) {
			return tokens.get(i).getLine();
		}
		else if (eofToken != null) {
			return eofToken.getLine();
		}
		else if (tokens.size() > 0) {
			// have to calculate the result from the line/column of the previous
			// token, along with the text of the token.
			Token lastToken = tokens.get(tokens.size() - 1);
			int line = lastToken.getLine();

			String tokenText = lastToken.getText();
			if (tokenText != null) {
				for (int i = 0; i < tokenText.length(); i++) {
					if (tokenText.charAt(i) == '\n') {
						line++;
					}
				}
			}

			// if no text is available, assume the token did not contain any newline characters.
			return line;
		}

		// only reach this if tokens is empty, meaning EOF occurs at the first
		// position in the input
		return 1;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public CharStream getInputStream() {
		if (i < tokens.size()) {
			return tokens.get(i).getInputStream();
		}
		else if (eofToken != null) {
			return eofToken.getInputStream();
		}
		else if (tokens.size() > 0) {
			return tokens.get(tokens.size() - 1).getInputStream();
		}

		// no input stream information is available
		return null;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String getSourceName() {
		if (sourceName != null) {
			return sourceName;
		}

		CharStream inputStream = getInputStream();
		if (inputStream != null) {
			return inputStream.getSourceName();
		}

		return "List";
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public void setTokenFactory(TokenFactory<?> factory) {
		this._factory = factory;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public TokenFactory<?> getTokenFactory() {
		return _factory;
	}
}
